ambry-router/src/test/java/com.github.ambry.router/GetOperationTest.java

```diff
@@ -0,0 +1,474 @@
+            operationCallback, time);
+    ArrayList<RequestInfo> requestListToFill = new ArrayList<>();
+    requestRegistrationCallback.requestListToFill = requestListToFill;
+
+    for (MockServer server : mockServerLayout.getMockServers()) {
+      server.setServerErrorForAllRequests(ServerErrorCode.Blob_Not_Found);
+    }
+
+    while (!op.isOperationComplete()) {
+      op.poll(requestRegistrationCallback);
+      List<ResponseInfo> responses = sendAndWaitForResponses(requestListToFill);
+      for (ResponseInfo response : responses) {
+        op.handleResponse(response);
+        if (op.isOperationComplete()) {
+          break;
+        }
+      }
+    }
+
+    Assert.assertEquals("Must have attempted sending requests to all replicas", replicasCount,
+        correlationIdToGetOperation.size());
+    Assert.assertTrue("Operation should be complete at this time", op.isOperationComplete());
+    Assert.assertTrue(operationFuture.isDone());
+    RouterException routerException = (RouterException) operationException.get();
+    Assert.assertEquals(RouterErrorCode.BlobDoesNotExist, routerException.getErrorCode());
+  }
+
+  /**
+   * Test the case with Blob_Not_Found errors from most servers, and Blob_Deleted at just one server. The latter
+   * should be the exception received for the operation.
+   * @throws Exception
+   */
+  @Test
+  public void testErrorPrecedenceWithBlobDeletedCase()
+      throws Exception {
+    GetBlobInfoOperation op =
+        new GetBlobInfoOperation(routerConfig, mockClusterMap, responseHandler, blobIdStr, operationFuture,
+            operationCallback, time);
+    ArrayList<RequestInfo> requestListToFill = new ArrayList<>();
+    requestRegistrationCallback.requestListToFill = requestListToFill;
+
+    int i = 0;
+    int indexToSetBlobDeleted = random.nextInt(replicasCount);
+    for (MockServer server : mockServerLayout.getMockServers()) {
+      if (i == indexToSetBlobDeleted) {
+        server.setServerErrorForAllRequests(ServerErrorCode.Blob_Deleted);
+      } else {
+        server.setServerErrorForAllRequests(ServerErrorCode.Blob_Not_Found);
+      }
+      i++;
+    }
+
+    while (!op.isOperationComplete()) {
+      op.poll(requestRegistrationCallback);
+      List<ResponseInfo> responses = sendAndWaitForResponses(requestListToFill);
+      for (ResponseInfo response : responses) {
+        op.handleResponse(response);
+        if (op.isOperationComplete()) {
+          break;
+        }
+      }
+    }
+
+    Assert.assertTrue("Operation should be complete at this time", op.isOperationComplete());
+    Assert.assertTrue(operationFuture.isDone());
+    RouterException routerException = (RouterException) operationException.get();
+    Assert.assertEquals(RouterErrorCode.BlobDeleted, routerException.getErrorCode());
+  }
+
+  /**
+   * Test the case with Blob_Not_Found errors from most servers, and Blob_Expired at just one server. The latter
```

Comment:
suggestion: can reduce code (and make writing new tests easier) by combining this test and the previous one.

For now you can just put all this code inside a for loop that loops on a list that contains Deleted and Expired, but I would be all for a general test that takes params like

`Map<ServerErrorCode, Integer> serverErrorCodeCounts, RouterErrorCode expectedErrorCode`.

and another with

`ServerErrorCode[] serverErrorCodesInOrder, RouterErrorCode expectedErrorCode`.

We will have a check to ensure that the counts add up to the number of servers (if count < noServers, the rest return `NO_ERROR`, if count > noServers, we throw). 

I believe this can be more of a Test Utility. All of put, get, delete and any other operation will simply use this function to check for precedence and they can mix and match however they want. Adding new tests would be easy.

Ming already has some tests in delete manager for checking precedence and I took a quick look at them and it seems like that test code can be greatly reduced and simplified if we have these functions.